### ISSUE TITLE
root.readOnly: true does not work on filesystems mounted with nodev or nosuid

### ISSUE THREAD

**Issue by smaeul** [opencontainers/runc#1247]
I'm using runc 1.0.0_rc2 on Linux 4.7.10 with grsecurity. Say I have a container rootfs mounted like so:
`rpool/srv/test/rootfs-1 on /srv/test/rootfs type zfs (rw,nosuid,nodev,noatime,xattr,noacl)`
I have in my /srv/test/config.json: `"root": {
    "path": "rootfs",
    "readonly": true
}`

When I run `runc run -b /srv/test test`, I get `container_linux.go:247: starting container process caused "process_linux.go:359: container init caused \"rootfs_linux.go:110: setting rootfs as readonly caused \\\"operation not permitted\\\"\""`

strace tells me the failing mount call is
`mount("/", "/", 0xc4200da900, MS_RDONLY|MS_REMOUNT|MS_BIND|MS_REC, NULL) = -1 EPERM`

This looks to be from [setReadonly()](https://github.com/opencontainers/runc/blob/f376b8033d2caae5207de9be1ee4d4a695886af1/libcontainer/rootfs_linux.go#L602) manually specifying the mount flags and not respecting existing ones.

If I remount the rootfs like so:
`host0/srv/test/rootfs-1 on /srv/test/rootfs type zfs (rw,noatime,xattr,noacl)`
starting the container succeeds.

This can be fixed by changing setReadonly() to also pass MS_NODEV/MS_NOEXEC/MS_NOSUID to mount() if they are already present in the mount options.

**Comment by cyphar** [comment]
@justincormack just changed the code to be the way it is today (#1222) , because `MS_REMOUNT` doesn't retain some flags properly. WDYT @justincormack?

**Comment by mrunalp** [comment]
We saw similar issue when trying to bind mount a directory which is a mount. We can read the existing mount options and pass them on remount. I am working on a fix for it.

**Comment by justincormack** [comment]
Hmm, I don't think you need to read the existing mount options, it should be possibly to do a bind again if this fails and then remount read only. Will take a look at this case.

**Comment by rhatdan** [comment]
We are now seeing this issue in testing and running a locked down crio instance.

**Comment by AkihiroSuda** [comment]
@brauner is working on the kernel-side fix for recursive read-only mount https://github.com/moby/moby/issues/37838

**Comment by cyphar** [comment]
Right, I know. I talked to him about it a while ago (he's currently busy with a few other projects though).

**Comment by rhatdan** [comment]
My understanding of MS_REC is ONLY for setting the way "sharing" is handled.  No other mount options are handled with it.  SHARED,RSHARED, PRIVATE,RPRIVATE, SLAVE, RSLAVE.

**Comment by cyphar** [comment]
`MS_REC` is used for bind-mounts.

**Comment by rhatdan** [comment]
Right but it only sets the propogation on bind-mounts not the read-only option.

**Comment by cyphar** [comment]
I think we're talking past each other -- `MS_REC|MS_BIND` allows for recursive bind-mounts which is quite different to just setting propagation. So `MS_REC` is used for things other than sharing.

You're quite right that `MS_REC|MS_RDONLY` doesn't do anything at the moment -- my point is that it probably should do what most users would expect it to do (recursively make things read-only). In the meantime we'll have to do the same looping over `/proc/mountinfo` ourselves...

**Comment by rhatdan** [comment]
Yes, I guess we are in agreement.  I have failed at least twice making this assumption, and have had to jump through hoops to get the ro bindmount to work in code.  Podman has a hacky version to bind mount  the cgroup hierarchy read/only into a container
